Add controllers

lazarx77 opened the pull request:
Доработка функционала согласно ТЗ спринта 14.

lazarx77 committed:
feature/ user controller created

lazarx77 committed:
feature/ item controller created

lazarx77 committed:
fix/ readme fix

lazarx77 committed:
fix/ readme fix

kriwsha commented on README.md:
```diff
@@ -1,2 +1,3 @@
 # java-shareit
-Template repository for Shareit project.
+Доработка функционала согласно ТЗ спринта 14.
```
Вообще, функционал != функциональность. Функционал, простыми словами, - это функция от функции. А реализуем мы функциональности. Но да, "функционал" вы будете слышать чаще, хотя это неправильно))

lazarx77 replied:
Меня и самого с самого начала размещения пулл-реквестов на гитхабе это коробило, но я делал, как было написано в примере. Поправил :-)

kriwsha commented on src/main/java/ru/practicum/shareit/exception/ItemDoNotBelongToUser.java:
```diff
@@ -0,0 +1,17 @@
+package ru.practicum.shareit.exception;
+
+/**
+ * Исключение ItemDoNotBelongToUser представляет собой пользовательское исключение,
```
Суперподробное описание - это и хорошо, и плохо. С одной стороны, вы хорошо и понятно объяснили. С другой стороны, вряд ли кто это дочитает до конца. Оставляю на ваш выбор, но на реальных проектах лучше писать кратко (я бы сказал, лапидарно=) ). Н-р, "Исключение для случая, когда предметом пытается управлять тот, кому предмет не принадлежит". То, что это RuntimeException, видно из сигнатуры класса (об этом можно в комментарии не писать).

lazarx77 replied:
На самом деле я не знаю, когда вовремя остановиться :-). Учитывая, что я осознаю, что это учебный проект, я пробую разные варианты комментариев, чтобы посмотреть, как это все работает. Я пробежался по всему проекту и убрал избыточную (на мой взгляд) информацию из комментариев.

kriwsha commented on src/main/java/ru/practicum/shareit/item/InMemoryItemRepository.java:
```diff
@@ -0,0 +1,99 @@
+import java.util.Optional;
+
+/**
+ * Класс InMemoryItemRepository реализует интерфейс ItemRepository и предоставляет
```
Здесь тоже можно сократить. В комментарии лучше писать назначение без технической реализации: то, что класс реализует интерфейс, видно из сигнатуры класса.

lazarx77 replied:
Хорошо. Пока немного не понимаю, когда вовремя остановиться. Спасибо за разъяснения :-).

kriwsha commented on src/main/java/ru/practicum/shareit/item/ItemService.java:
```diff
@@ -0,0 +1,78 @@
+ * <p>Методы интерфейса:</p>
+ * <ul>
+ *     <li>
+ *         {@link #addNewItem(Long, ItemDto)} - добавляет новый предмет в систему.
```
Описание методов лучше оставить нам самими методами

lazarx77 replied:
Поправил комментарии по всему проекту - убрал избыточное описание методов из "шапки" классов :-).

kriwsha commented on src/main/java/ru/practicum/shareit/item/ItemValidatorService.java:
```diff
@@ -0,0 +1,31 @@
+ * и аннотацию {@link Slf4j} для логирования.</p>
+ */
+@Slf4j
+@Service
```
У вас классе один статический метод. Класс больше похож на утилитарный класс. Доступаетесь вы до него не через создание бина, а через имя класса. Поэтому аннотация Service здесь не нужна.

lazarx77 replied:
Действительно, я об этом не подумал. Видимо, аннотацию Service поставил на автомате, так как класс называется тоже "...service". Поправил :-)

kriwsha commented on src/main/java/ru/practicum/shareit/user/UserValidatorService.java:
```diff
@@ -0,0 +1,46 @@
+     */
+    public static void validateEmailDouble(User user, List<User> users) {
+        for (User user1 : users) {
+            if (user1.equals(user)) { //@EqualsAndHashCode(of = "email")
```
Закомментированный код лучше удалять. И еще я бы все таки явно сравнивал емайлы. Без комментария логика неочевидна, что может вызвать путаницу.

lazarx77 replied:
Изначально сделал так, так как было интересно посмотреть, а "сработает ли" такая реализация. Комментарий убрал, прописал явную проверку email пользователей, внес изменения в методы обновления пользователей в UserServiceImpl и InMemoryUserRepository.  Также убрал ненужную аннотацию Service у класса.

kriwsha commented on src/main/java/ru/practicum/shareit/user/UserValidatorService.java:
```diff
@@ -0,0 +1,46 @@
+     * @throws EmailDoubleException если электронная почта уже используется другим пользователем.
+     */
+    public static void validateEmailDouble(User user, List<User> users) {
+        for (User user1 : users) {
```
Я стараюсь не называть переменные с цифрами, если этого не требуется явно. Н-р, если нужно создать переменную, которая хранит строку "ГОСТ Р ИСО/МЭК 20546-2021", то да, тут ее можно назвать GOST_20546_2021. Но когда у нас две переменные с одинаковыми названиями, то я называю либо "sourceUser <-> targetUser", либо "sourceUser <-> currentUser". Но это только рекомендация.

lazarx77 replied:
Согласен, мне sourceUser и currentUser тоже нравятся больше. Поправил.

Как всегда, большое спасибо за полезные рекомендации!👍🤝

lazarx77 committed:
fix/ - readme fix,
- documented comments shrunk - unnecessary methods descriptions removed,
- UserValidatorService and ItemValidatorService Service annotation removed,
- in UserValidatorService validateEmailDouble method updated to compare users emails explicitly,
- UserServiceImpl and InMemoryUserRepository update methods updated.
